benchexec/resources.py

```diff
@@ -15,191 +15,199 @@
-        allCpus = my_cgroups.read_allowed_cpus()
-
-        # Filter CPU cores according to the list of identifiers provided by a user
-        if coreSet:
-            invalid_cores = sorted(set(coreSet).difference(set(allCpus)))
-            if len(invalid_cores) > 0:
-                raise ValueError(
-                    "The following provided CPU cores are not available: "
-                    + ", ".join(map(str, invalid_cores))
-                )
-            allCpus = [core for core in allCpus if core in coreSet]
+        # read list of available CPU cores (int)
+        allCpus_list = get_cpu_list(my_cgroups, coreSet)
 
-        logging.debug("List of available CPU cores is %s.", allCpus)
+        # read & prepare hyper-threading information, filter redundant entries
+        siblings_of_core = get_siblings_mapping(allCpus_list)
+        cleanList = []
+        for core in siblings_of_core:
+            if core not in cleanList:
+                for sibling in siblings_of_core[core]:
+                    if sibling != core:
+                        cleanList.append(sibling)
+        for element in cleanList:
+            siblings_of_core.pop(element)
+        # siblings_of_core will be added to hierarchy_levels list after sorting
 
-        # read mapping of core to memory region
-        cores_of_memory_region = collections.defaultdict(list)
-        for core in allCpus:
-            coreDir = f"/sys/devices/system/cpu/cpu{core}/"
-            memory_regions = _get_memory_banks_listed_in_dir(coreDir)
-            if memory_regions:
-                cores_of_memory_region[memory_regions[0]].append(core)
-            else:
-                # If some cores do not have NUMA information, skip using it completely
-                logging.warning(
-                    "Kernel does not have NUMA support. Use benchexec at your own risk."
-                )
-                cores_of_memory_region = {}
-                break
-        logging.debug("Memory regions of cores are %s.", cores_of_memory_region)
-
-        # read mapping of core to CPU ("physical package")
-        cores_of_package = collections.defaultdict(list)
-        for core in allCpus:
-            package = get_cpu_package_for_core(core)
-            cores_of_package[package].append(core)
-        logging.debug("Physical packages of cores are %s.", cores_of_package)
-
-        # select the more fine grained division among memory regions and physical package
-        if len(cores_of_memory_region) >= len(cores_of_package):
-            cores_of_unit = cores_of_memory_region
-            logging.debug("Using memory regions as the basis for cpu core division")
-        else:
-            cores_of_unit = cores_of_package
-            logging.debug("Using physical packages as the basis for cpu core division")
+        # read & prepare mapping of cores to L3 cache
+        cores_of_L3cache = get_L3cache_mapping(allCpus_list)
+        hierarchy_levels.append(cores_of_L3cache)
```

Comment:
Why are in some cases the return values of these methods checked, and in others not?

Reply:
The thought behind that was that some methods may return empty lists if not supported by the processor (e.g. NUMA) while there should always be a socket or a level 3 cache (even if it is not shared). 
This is changed in commit https://github.com/sosy-lab/benchexec/pull/892/commits/4b5c08435488c589cb62875273ce834abb56d688 : now all return values are checked.